### PR TITLE
fix(host-service): refresh minted JWTs after source token rotation

### PR DESCRIPTION
**Pull request by DivyamTalwar**
## Summary

This PR fixes host-service JWT cache reuse after the underlying API key or session token rotates.

## Why this matters

`JwtApiAuthProvider` cached exchanged JWTs only by time. For non-JWT credentials (API keys or session tokens), that meant a host could continue sending a JWT minted from the previous source credential until the 55-minute cache window expired, even after the credential source started returning a new token.

That is production-relevant for login refreshes, API key rotation, and revoked/replaced credentials because host→cloud calls can keep authenticating with stale credential-derived JWTs longer than intended.

## Changes

- Reads the current source token before deciding whether an exchanged JWT cache entry is reusable.
- Tracks which source token produced the cached JWT.
- Reuses a cached exchanged JWT only when it is still fresh **and** was minted from the same current source token.
- Clears the cached source-token binding on invalidation.
- Adds regression coverage for API key rotation without explicit invalidation.

## Validation

Commands run:

- `bun --cwd packages/host-service test src/providers/auth/JwtAuthProvider/JwtAuthProvider.test.ts` — passed
- `bunx @biomejs/biome@2.4.2 check packages/host-service/src/providers/auth/JwtAuthProvider/JwtAuthProvider.ts packages/host-service/src/providers/auth/JwtAuthProvider/JwtAuthProvider.test.ts` — passed
- `git diff --check` — passed

Failures:

- `bun --cwd packages/host-service typecheck` — failed
  - Related to this PR: no
  - Evidence: existing failure in `src/trpc/router/git/get-check-job-logs.test.ts(52,42)` where a plain `string` is passed to a TRPC error-code matcher expecting the TRPC error-code union.
  - Follow-up: should be handled separately because this PR only touches auth-provider cache behavior.

## Risk

Risk level: low

Compatibility impact: no public API changes. Stable source tokens still reuse cached exchanged JWTs. Direct OAuth JWT source tokens still pass through without exchange.
Rollback simplicity: revert the provider cache-binding change and the regression test.
Remaining edge cases: this intentionally does not remove JWT caching; it only prevents reuse across source credential rotations.

## Issue

No existing issue. This was discovered during repository analysis.

## Notes for maintainers

The fix preserves the existing invalidation path and cache window. It only adds source-token identity as a cache key for exchanged JWTs.

---

<a href="https://stagereview.app/superset-sh/superset/pull/5473">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

---
## Summary by cubic
Bind cached exchanged JWTs in `packages/host-service` to the current source token so hosts refresh immediately after API key or session token rotation. This prevents stale host→cloud auth while keeping direct OAuth JWT passthrough unchanged.

- **Bug Fixes**
  - Reuse a cached exchanged JWT only if it was minted from the current source token and is still fresh.
  - Read the latest source token before reuse; invalidate clears both the JWT and its source-token binding.
  - Added a test verifying a new JWT is minted after API key rotation.

<sup>Written for commit 82a09768f018d2b2400a6941d3018f059d671f67. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/superset-sh/superset/pull/5473?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed JWT reuse after API key rotation. The app now refreshes the cached JWT when the underlying source key changes, preventing stale authentication headers.
  * Added coverage to verify the old token is replaced with a newly minted one after rotation.

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 3 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Track source token for cached JWTs](https://stagereview.app/superset-sh/superset/pull/5473/chapters/1) |
| 2 | [Validate source token identity during cache lookup](https://stagereview.app/superset-sh/superset/pull/5473/chapters/2) |
| 3 | [Verify cache refresh on API key rotation](https://stagereview.app/superset-sh/superset/pull/5473/chapters/3) |

<a href="https://stagereview.app/superset-sh/superset/pull/5473">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 8831f9ea628a6b48c80586dfefea7d33bd100e15 on Jul 6, 2026 6:54pm UTC.</sub>

**Comment by DivyamTalwar** [comment]
### Keyless before/after verification

`JwtApiAuthProvider` cached the minted JWT on a TTL only, never tracking which source token produced it — so after the source key/token rotates, `getJwt()` kept returning the **stale** JWT for the whole 5-minute window. Fix keys the cache on the source token.

summary

**Before — base `main`**  
Base code presents `Authorization: Bearer minted.old.jwt` after rotation → **1 fail**.

before

**After — the fix**  
Fresh JWT minted after rotation → **9 pass / 0 fail**.

after

_Verification is fully keyless — `bun test` with mocked fetch/DB/auth, no API key or cloud login. Raw transcripts + full write-up are in the linked evidence branch._

**CI:** CodeRabbit ✅ pass · cubic skipped · `mergeable = MERGEABLE`, single clean commit. `BLOCKED` = awaiting required maintainer review — no failing pipeline checks (the repo's build/test workflows don't run on fork PRs). PR-owned files pass Biome 2.4.2.

@Kitenite — small, self-contained **host-service auth** fix with a focused regression test; would appreciate a review when you have a moment. 🙏

File: packages/host-service/src/providers/auth/JwtAuthProvider/JwtAuthProvider.test.ts [hunk 1]
```diff
@@ -357,6 +357,38 @@ describe("JwtApiAuthProvider with config-backed host auth", () => {
 		}
 	});
 
+	test("does not reuse a minted JWT after the source API key rotates", async () => {
```
[the remark this answers is not in the document]
Comment (DivyamTalwar) [reply packages/host-service/src/providers/auth/JwtAuthProvider/JwtAuthProvider.test.ts:381]: Moved the rotation test into its own top-level `describe("JwtApiAuthProvider source-token rotation")` (82a09768f). Thanks!

File: packages/host-service/src/providers/auth/JwtAuthProvider/JwtAuthProvider.ts [hunk 2]
```diff
@@ -63,6 +58,14 @@ export class JwtApiAuthProvider implements ApiAuthProvider {
 			return sessionToken;
 		}
 
+		if (
+			this.cachedJwt &&
+			this.cachedJwtSourceToken === sessionToken &&
+			Date.now() < this.cachedJwtExpiresAt - JWT_REFRESH_BUFFER_MS
+		) {
+			return this.cachedJwt;
+		}
```
[the remark this answers is not in the document]
Comment (DivyamTalwar) [reply packages/host-service/src/providers/auth/JwtAuthProvider/JwtAuthProvider.ts:67]: Intentional: detecting a rotated source token *requires* reading the current token on every call — that's the crux of the fix. The cache check now runs after `getSessionToken()`, so a rotated key invalidates the cached JWT immediately instead of serving a stale one for up to 5 minutes. The source here is a local config-file read (`ConfigFileSessionTokenSource`), so the added cost is negligible relative to correctly honoring rotation. Resolving as by-design.
